keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,565 @@
+container consuming it has started.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+The user is using a container offered by a vendors that requires configuring
+env vars. [Some
+vendors](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/examples#environment-variable-example)
+have to couple their app with Kubernetes to allow setting env vars in their
+container.
+
+Instead, k8s can provide a new mechanism that will be used to instantiate
+env vars from the specified file. The user can generate an env file using an
+`initContainer`. The container can then use the `fileRef` field to reference
+the file from which it should initialize the env vars. Behind the scene, kubelet will
+parse the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "CONFIG_VAR=HELLO" > /etc/config/config.env']
+    volumeMounts:
+    - name: data
+      mountPath: /etc/config
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    envFrom:
+    - fileRef:
+        path: config.env
+        volumeName: data
+  restartPolicy: Never
+  volumes:
+  - name: data
+    emptyDir: {}
+```
+
+#### Story 2
+
+Each k8s Node is instantiated with an env file that contains information about
+the Node (for eg, instance type, failure zones). The user would like to
+provide some of the information in this file to the vendor's container via
+env vars. For example, consider that each node has the following env file:
+
+```
+CONFIG_VAR=HELLO
+CONFIG_VAR_A=WORLD
+...
+```
+
+The user can use the `fileKeyRef` field to select a subset of these env vars
+that it wants to provide to the app. Behind the scene, kubelet will parse the
+env file on the host and instantiate these env vars when creating the main
+container. It does not mount this hostPath volume onto the main container.
+In this case, the Pod's output includes `CONFIG_VAR=HELLO`
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'cp /etc/config/config.env /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+    - name: data
+      mountPath: /etc/config
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+  - name: data
+    hostPath: /etc/config
```

Comment:
Can we modify the example to not include hostPath anymore?

Reply:
Removed the hostPath example.